clang/lib/Sema/TreeTransform.h

```diff
@@ -7081,10 +7085,10 @@ QualType TreeTransform<Derived>::TransformAttributedType(
   // FIXME: dependent operand expressions?
   if (getDerived().AlwaysRebuild() ||
       modifiedType != oldType->getModifiedType()) {
-    // TODO: this is really lame; we should really be rebuilding the
-    // equivalent type from first principles.
-    QualType equivalentType
-      = getDerived().TransformType(oldType->getEquivalentType());
+    TypeLocBuilder AuxiliaryTLB;
+    AuxiliaryTLB.reserve(TL.getFullDataSize());
```

Comment:
an alternative might be to call 
`TLB.TypeWasModifiedSafely(result)` (after the call to `getAttributedType` below).
I wonder if @AaronBallman  has a better alternative

Reply:
Right, yeah, the location shouldn't have changed, just the type itself.  I think the suggestions above are the way to go.

Reply:
Thanks for your remind! Fixed.

Reply:
I don't think we need to use `AuxiliaryTLB` any longer, do we?

Reply:
I‘m afraid not. It will crash on `TypeLocBuilder::pushImpl` with the new added test case since `TLast == LastTy` assert failed if we reuse `TLB`. Because current `FunctionProtoType` can't be child of other type node.

Reply:
Ah, you're right, we do still need to do that dance it seems.

Reply:
@cor3ntin @AaronBallman @erichkeane The test on windows platform failed with `TLB.TypeWasModifiedSafely(result);`. Maybe we shouldn't change `LastTy` of `TLB` since we just transform `EquivalentType` using `AuxiliaryTLB` and haven't touched `TLB`.

Reply:
I tested `address-spaces.cpp` on linux and `TLast == LastTy` assert also failed. Don't know why the test on linux platform passed.

Reply:
Yeah, I was thinking we didn't need to use `AuxiliaryTLB` and could instead call `TypeWasModifiedSafely` but it seems like we need the code as you originally had it (using `AuxiliaryTLB` but not calling `TypeWasModifiedSafely`). Sorry for the run-around!